`PEImage` should not permit `m_path` field mutation

AaronRobinsonMSFT opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/90657

Remove cases where `PEImage::m_path` was mutated.
Create `m_pathHash` field and remove function.
Remove `FEATURE_CASE_SENSITIVE_FILESYSTEM`.

jkotas commented on src/coreclr/inc/sstring.h:
```diff
@@ -125,8 +125,7 @@ class EMPTY_BASES_DECL SString : private SBuffer
 
     SString();
-
-    explicit SString(const SString &s);
```
What are all existing places that use this constructor? Can they be broken by changing what this constructor does?

Are there places that will unexpectedly start using this constructor with removal of `explicit` (and may get broken)?

I am wondering whether it would be better to make this explicit so one can tell by looking at the callsites what is exactly going on.

AaronRobinsonMSFT replied:
> Can they be broken by changing what this constructor does?

Yes they can. It is a huge mess to detangle this in general. That is why this PR is draft. I wanted to see across systems quickly how this behaves. I am still looking into approaches that will make this work in an intuitive way.

AaronRobinsonMSFT replied:
I've convert to move ctor/assignment.

jkotas commented on src/coreclr/vm/peimage.inl:
```diff
@@ -385,24 +389,6 @@ inline BOOL PEImage::HasPath()
     return !GetPath().IsEmpty();
 }
```
Delete `HasPath`? Is it no longer used and it is inefficient now.

AaronRobinsonMSFT replied:
Oops. I thought I reverted that. Thanks.

AaronRobinsonMSFT committed:
Remove cases where PEImage::m_path was mutated.
Create m_pathHash field and remove function.
Remove FEATURE_CASE_SENSITIVE_FILESYSTEM.

jkotas commented on src/coreclr/vm/peimage.cpp:
```diff
@@ -141,11 +141,14 @@ ULONG PEImage::Release()
         {
-            LOG((LF_LOADER, LL_INFO100, "PEImage: Closing Image %s\n", m_path.GetUTF8()));
+#ifdef LOGGING
+            SString path{ m_path };
```
This is NOTHROW method. Is this going to hit contract violation? (The problem was there even before this change, but it is going to happen in more situations now.)

AaronRobinsonMSFT committed:
Address contract violations

jkotas commented on src/coreclr/vm/dwbucketmanager.hpp:
```diff
@@ -960,11 +960,18 @@ bool BaseBucketParamsManager::GetFileVersionInfoForModule(Module* pModule, USHOR
+            EX_TRY
             {
-                succeeded = true;
+                SString modulePath{ pPEAssembly->GetPath() };
```
This is executing during watson handling. Introducing new allocations in the watson handling is potentially problematic. (I am not sure how much we allocate during watson handling today, but I would expect we try to minimize it.)

I would keep this as is, and maybe just assert that the string we got has `IsNormalized()` set and thus `GetUnicode` won't allocate or cause conversions.

jkotas reviewed (approved):
LGTM otherwise. Thank you!

AaronRobinsonMSFT committed:
Remove try/catch and replace with assert

AaronRobinsonMSFT commented:
tvos failure is unrelated.
